gramps/gen/types.py

```diff
@@ -84,3 +110,194 @@
     NoteGrampsID,
 ]
 AnyGrampsID = PrimaryObjectGrampsID  # No Tag IDs
+
+
+@runtime_checkable
+class GrampsTypeLike(Protocol):
+    """Protocol for GrampsType-like objects (NameType, AttributeType, etc.)."""
```

Comment:
If protocols are used, they should be the 'primary' types. In other words, concrete types should refer to the protocols, and not the other way around, so a `GrampsType` is defined as a `GrampsTypeLike`, but a `GrampsTypeLike` is not defined as a `GrampsType`-like.

Reply:
I don't agree actually - unless I'm misunderstanding what you mean. The whole point of protocols is that they can be used without inheritance. For instance, `DataDict` cannot inherit from `PersonLike` because not every `DataDict` has a `surname_list` (it might be a family or an event etc.).

It's like with database proxies in Gramps that sometimes don't inherit from `DbGeneric` but just share a subset of its methods.

Reply:
If protocols are used like interfaces (as in the abstract programming concepts), such interfaces should not reference concrete types. For example, a 'person-like' interface is not a 'person' (concrete type). Instead, it's the other way around: a 'person' concrete type is a 'person-like' interface. That way other code can type on the interface (the protocol, in Python terminology), instead of any concrete type.

Reply:
I agree. Anything in the code or in what I said that goes against that?

Reply:
I'd be happy to complete it for all the other primary types so we can make use of it in other PRs if the feedback is positive.